service/storage_proxy.cc

```diff
@@ -1348,14 +1348,22 @@ endpoints_to_replica_ids(const locator::token_metadata& tm, const inet_address_v
 }
 
 query::max_result_size storage_proxy::get_max_result_size(const query::partition_slice& slice) const {
-    // Unpaged and reverse queries.
-    // FIXME: some reversed queries are no longer unlimited.
-    // To filter these out we need to know if it's a single partition query.
-    // Take the partition key range as a parameter? This would require changing many call sites
-    // where the partition key range is not immediately available.
-    // For now we ignore the issue, return the 'wrong' limit for some reversed queries, and wait until all reversed queries are fixed.
-    if (!slice.options.contains<query::partition_slice::option::allow_short_read>() || slice.options.contains<query::partition_slice::option::reversed>()) {
+    if (!slice.options.contains<query::partition_slice::option::allow_short_read>()) { // Unpaged
         return _db.local().get_unlimited_query_max_result_size();
+    } else if (slice.options.contains<query::partition_slice::option::reversed>()) { // Reversed queries
+        if (_features.cluster_supports_reduced_revenced_queries_limit()) {
+            // If the whole cluster supports new more efficient reversed
+            // queries, then we can use regular limit. We still need to set
+            // unlimited_* limits for reversing_reader that's still used for
+            // KA/LA sstables.
+            auto limit = _db.local().get_unlimited_query_max_result_size();
+            limit.soft_limit = query::result_memory_limiter::maximum_result_size;
+            limit.hard_limit = query::result_memory_limiter::maximum_result_size;
+            return limit;
+        } else {
+            return _db.local().get_unlimited_query_max_result_size();
+        }
```

Comment:
With the new interface, the implementation should look like this:
```c++
if (_features.cluster_supports_reduced_revenced_queries_limit()) {
     return {_db.local().get_unlimited_query_max_result_size(), query::result_memory_limiter::maximum_result_size};
}
// fall back to old logic
```
We can do this because now the replicas get both the normal page size as well as the hard/soft limit for any activity with unbounded memory usage. The coordinator need not care about what query this is, the replica will know to use the right limit for the right job.

Reply:
Could you explain what the `{_db.local().get_unlimited_query_max_result_size(), query::result_memory_limiter::maximum_result_size}` constructor does in the solution you're proposing? At the moment it would set soft limit to `_db.local().get_unlimited_query_max_result_size()` and hard limit to `query::result_memory_limiter::maximum_result_size` which makes no sense.

Reply:
What I really mean is:
```
const auto& limits = _db.local().get_unlimited_query_max_result_size();
return max_result_size{.soft_limit = limits.soft_limit, .hard_limit = limits.hard_limit, .page_size = query::result_memory_limiter::maximum_result_size};
```
Basically, we don't need logic on which fields to fill/not to fill, we always fill them with the relevant values.